hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/CompositeRecordStoreMutationObserver.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.map.impl.recordstore;
+
+import com.hazelcast.map.impl.record.Record;
+import com.hazelcast.nio.serialization.Data;
+
+import java.util.Collection;
+import java.util.LinkedList;
+
+class CompositeRecordStoreMutationObserver<R extends Record> implements RecordStoreMutationObserver<R> {
```

Comment:
Just a little thing to consider: should we trust the contained observers, if one of the observers fails with an exception should we continue invoking other observers?

Reply:
This is a very good point. The original code didn't have any special exception handling. If the `EventJournal` call failed, it failed the operation and the subsequent tasks were not executed (map event publishing, TTL setting, index updates, `storage.remove()` etc). 

Since we have no transactional behavior here, an exception leaves inconsistency behind. Either in the part that the failing observer called (swallowing the exception) or in all subsequent tasks (throwing immediately). There is a third option with caching the exception, continuing with the other observers and throwing the exception in the end. I'm not sure if any of the three is a better option than the other two, the result is inconsistency. I see this as a general problem that the map operations are not atomic. 

Also, as the purpose of this abstraction and the composite impl is to hide complexity from the `RecordStore` and to decouple things, I'd keep it as it is.

Reply:
Regarding this, I took care to call the event journal AFTER the storage was updated (at least in most cases). This is the way the map store was called, which suffers from the same problem - it can fail even after the record store is updated.